### PR TITLE
Feature/cdap 1471 object table

### PR DESCRIPTION
**Pull request by albertshau**

### CODE REVIEW

**Comment by albertshau** [comment]
http://builds.cask.co/browse/CDAP-DUT803

File: cdap-api/src/main/java/co/cask/cdap/api/dataset/lib/ObjectMappedTable.java [hunk 1]
```diff
@@ -0,0 +1,104 @@
...
+public interface ObjectMappedTable<T> extends Dataset, BatchReadable<byte[], T>,
+  BatchWritable<byte[], T>, RecordScannable<StructuredRecord> {
+
+  /**
+   * Write an object with a given key.
+   *
+   * @param key the key of the object
+   * @param object the object to be stored
+   */
+  void write(String key, T object);
```
Comment (anew) [inline cdap-api/src/main/java/co/cask/cdap/api/dataset/lib/ObjectMappedTable.java]: should it be put and get? instead of write and read?

File: cdap-api/src/main/java/co/cask/cdap/api/dataset/lib/ObjectMappedTable.java [hunk 2]
```diff
@@ -0,0 +1,104 @@
...
+  T read(byte[] key);
+
+  /**
+   * Scans table.
+   *
+   * @param startRow start row inclusive. {@code null} means start from first row of the table
+   * @param stopRow stop row exclusive. {@code null} means scan all rows to the end of the table
+   * @return {@link CloseableIterator} over {@link KeyValue KeyValue&lt;byte[], T&gt;}
+   */
+  CloseableIterator<KeyValue<byte[], T>> scan(@Nullable String startRow, @Nullable String stopRow);
```
Comment (anew) [inline cdap-api/src/main/java/co/cask/cdap/api/dataset/lib/ObjectMappedTable.java]: As a side thought: This seems right as KeyValueTable returns a CloseableIterator, too. I am just wondering why Table is returning a Scanner? Should we change Table?

File: cdap-api/src/main/java/co/cask/cdap/api/dataset/lib/ObjectStore.java [hunk 3]
```diff
@@ -37,7 +37,9 @@
  *   <li>a static inner class of one of the above</li>
  * </ul>
  *</p>
- * Interfaces and not-static inner classes are not supported.
+ * Interfaces and not-static inner classes are not supported. An ObjectStore will serialize the entire object and
+ * store it in a single cell. See {@link ObjectMappedTable} if you want object fields to be mapped to their own
+ * cells.
```
Comment (anew) [inline cdap-api/src/main/java/co/cask/cdap/api/dataset/lib/ObjectStore.java]: columns

File: cdap-data-fabric/src/main/java/co/cask/cdap/data/runtime/DataSetServiceModules.java [hunk 4]
```diff
@@ -76,6 +77,7 @@
     INMEMORY_DATASET_MODULES.put("timePartitionedFileSet", new TimePartitionedFileSetModule());
     INMEMORY_DATASET_MODULES.put("partitionedFileSet", new PartitionedFileSetModule());
     INMEMORY_DATASET_MODULES.put("aclTable", new ACLTableModule());
+    INMEMORY_DATASET_MODULES.put("objectMappedTable", new ObjectMappedTableModule());
```
Comment (anew) [inline cdap-data-fabric/src/main/java/co/cask/cdap/data/runtime/DataSetServiceModules.java]: yet another module... we need to fix that. Unrelated to this PR, though.

File: cdap-api/src/main/java/co/cask/cdap/api/dataset/lib/ObjectStores.java [hunk 5]
```diff
@@ -32,14 +30,39 @@
...
+   * @param datasetName dataset name
+   * @param type type of objects to be stored in {@link ObjectMappedTable}
+   * @param props any additional dataset properties
+   * @throws UnsupportedTypeException
+   */
+  public static void createObjectMappedTable(ApplicationConfigurer configurer,
+                                             String datasetName, Type type, DatasetProperties props)
+    throws UnsupportedTypeException {
+
+    configurer.createDataset(datasetName, ObjectMappedTable.class, objectStoreProperties(type, props));
```
Comment (anew) [inline cdap-api/src/main/java/co/cask/cdap/api/dataset/lib/ObjectStores.java]: So ObjectMappedTable uses the exact same properties as ObjectStore? But will that always be the case? Or should we have objectMappedTableProperties(...) for this?

File: cdap-data-fabric/src/main/java/co/cask/cdap/data2/dataset2/lib/table/ObjectMappedTableDefinition.java [hunk 6]
```diff
@@ -0,0 +1,104 @@
...
+
+  public ObjectMappedTableDefinition(String name, DatasetDefinition<? extends Table, ?> tableDef) {
+    super(name);
+    Preconditions.checkArgument(tableDef != null, "Table definition is required");
+    this.tableDef = tableDef;
+  }
+
+  @Override
+  public DatasetSpecification configure(String instanceName, DatasetProperties properties) {
+    Preconditions.checkArgument(properties.getProperties().containsKey("type"));
```
Comment (anew) [inline cdap-data-fabric/src/main/java/co/cask/cdap/data2/dataset2/lib/table/ObjectMappedTableDefinition.java]: So the properties must have type and schema? Why both?

File: cdap-data-fabric/src/main/java/co/cask/cdap/data2/dataset2/lib/table/ObjectMappedTableDataset.java [hunk 7]
```diff
@@ -0,0 +1,216 @@
...
+public class ObjectMappedTableDataset<T> extends AbstractDataset implements ObjectMappedTable<T> {
+
+  private final Table table;
+  private final Schema schema;
+
+  private final ReflectionPutWriter<T> putWriter;
+  private final ReflectionRowReader<T> rowReader;
+
+  @SuppressWarnings("unchecked")
+  public ObjectMappedTableDataset(String name, Table table, TypeRepresentation typeRep,
```
Comment (anew) [inline cdap-data-fabric/src/main/java/co/cask/cdap/data2/dataset2/lib/table/ObjectMappedTableDataset.java]: It is not clear to me why this needs both schema and type. Don't they both represent the same thing? Perhaps a Javadoc would be appropriate here.

File: cdap-data-fabric/src/test/java/co/cask/cdap/data2/dataset2/lib/table/Record.java [hunk 8]
```diff
@@ -0,0 +1,76 @@
...
+  public Record(int intField, Long longField, Float floatField, Double doubleField, String stringField,
+                byte[] byteArrayField, ByteBuffer byteBufferField, UUID uuidField) {
+    this.intField = intField;
+    this.longField = longField;
+    this.floatField = floatField;
+    this.doubleField = doubleField;
+    this.stringField = stringField;
+    this.byteArrayField = byteArrayField;
+    this.byteBufferField = byteBufferField;
+    this.uuidField = uuidField;
```
Comment (anew) [inline cdap-data-fabric/src/test/java/co/cask/cdap/data2/dataset2/lib/table/Record.java]: what is the schema for UUID?

File: cdap-data-fabric/src/main/java/co/cask/cdap/data2/dataset2/lib/table/ObjectMappedTableDataset.java [hunk 9]
```diff
@@ -0,0 +1,216 @@
...
+ *
+ * @param <T> the type of objects in the table
+ */
+@Beta
+public class ObjectMappedTableDataset<T> extends AbstractDataset implements ObjectMappedTable<T> {
+
+  private final Table table;
+  private final Schema schema;
+
+  private final ReflectionPutWriter<T> putWriter;
```
Comment (anew) [inline cdap-data-fabric/src/main/java/co/cask/cdap/data2/dataset2/lib/table/ObjectMappedTableDataset.java]: where did these get defined? They seem to be missing from the PR.

Comment (anew) [inline cdap-data-fabric/src/main/java/co/cask/cdap/data2/dataset2/lib/table/ObjectMappedTableDataset.java]: I guess I missed the previous PR for this. all good.

**Comment by anew** [comment]
Done with my first pass. Nothing  major, but please see my comments.

File: cdap-api/src/main/java/co/cask/cdap/api/dataset/lib/ObjectMappedTable.java (hunk 1 above)
Comment (albertshau) [inline cdap-api/src/main/java/co/cask/cdap/api/dataset/lib/ObjectMappedTable.java]: that would make it look more like the table api, which i think is better. will change.

File: cdap-api/src/main/java/co/cask/cdap/api/dataset/lib/ObjectMappedTable.java (hunk 2 above)
Comment (albertshau) [inline cdap-api/src/main/java/co/cask/cdap/api/dataset/lib/ObjectMappedTable.java]: i will open a jira

Comment (albertshau) [inline cdap-api/src/main/java/co/cask/cdap/api/dataset/lib/ObjectMappedTable.java]: CDAP-1547. I think it does make sense but should be done in a separate PR.

File: cdap-api/src/main/java/co/cask/cdap/api/dataset/lib/ObjectStore.java (hunk 3 above)
Comment (albertshau) [inline cdap-api/src/main/java/co/cask/cdap/api/dataset/lib/ObjectStore.java]: done

File: cdap-api/src/main/java/co/cask/cdap/api/dataset/lib/ObjectStores.java (hunk 5 above)
Comment (albertshau) [inline cdap-api/src/main/java/co/cask/cdap/api/dataset/lib/ObjectStores.java]: may diverge, I will create a separate method

File: cdap-data-fabric/src/test/java/co/cask/cdap/data2/dataset2/lib/table/Record.java (hunk 8 above)
Comment (albertshau) [inline cdap-data-fabric/src/test/java/co/cask/cdap/data2/dataset2/lib/table/Record.java]: UUID is treated as bytes by our reflection classes

File: cdap-data-fabric/src/main/java/co/cask/cdap/data2/dataset2/lib/table/ObjectMappedTableDataset.java (hunk 7 above)
Comment (albertshau) [inline cdap-data-fabric/src/main/java/co/cask/cdap/data2/dataset2/lib/table/ObjectMappedTableDataset.java]: we should be able to just take the type, will change

File: cdap-api/src/main/java/co/cask/cdap/api/dataset/lib/ObjectMappedTable.java (hunk 1 above)
Comment (albertshau) [inline cdap-api/src/main/java/co/cask/cdap/api/dataset/lib/ObjectMappedTable.java]: ok i see why, it's because BatchWritable uses ```write``` so I'm going to keep it as is (and remove the redundant ```write(byte[], T object)```. RecordWritable would also use ```write``` if we implement it.

File: cdap-api/src/main/java/co/cask/cdap/api/dataset/lib/ObjectStores.java (hunk 5 above)
Comment (albertshau) [inline cdap-api/src/main/java/co/cask/cdap/api/dataset/lib/ObjectStores.java]: actually, will create an ObjectMappedTableProperties class to be consistent with FileSetProperties and PartitionedFileSetProperties

File: cdap-data-fabric/src/main/java/co/cask/cdap/data2/dataset2/lib/table/ObjectMappedTableDefinition.java (hunk 6 above)
Comment (albertshau) [inline cdap-data-fabric/src/main/java/co/cask/cdap/data2/dataset2/lib/table/ObjectMappedTableDefinition.java]: want to validate the schema, and this is an earlier place than when somebody calls getDataset.  The schema cant be generated from the Type at this point because this gets called in the system context. If the Type is in the user code, it will not be available here.

Comment (albertshau) [inline cdap-data-fabric/src/main/java/co/cask/cdap/data2/dataset2/lib/table/ObjectMappedTableDefinition.java]: will add a comment about this

File: cdap-data-fabric/src/main/java/co/cask/cdap/data2/dataset2/lib/table/ObjectMappedTableDataset.java (hunk 7 above)
Comment (albertshau) [inline cdap-data-fabric/src/main/java/co/cask/cdap/data2/dataset2/lib/table/ObjectMappedTableDataset.java]: ok actually, because of the classloading issue described in the previous comment, since we have the schema in the definition it is simpler to just pass it in instead of generating it from the type each time a table is constructed

**Comment by albertshau** [comment]
@anew addressed comments, please take another look

File: cdap-api/src/main/java/co/cask/cdap/api/dataset/lib/ObjectMappedTable.java (hunk 1 above)
Comment (anew) [inline cdap-api/src/main/java/co/cask/cdap/api/dataset/lib/ObjectMappedTable.java]: That explains it. But wait.. this is BatchWritable<byte[], T>. So if at all, you would remove the write(String, T). But I think it is ok to keep both.

File: cdap-api/src/main/java/co/cask/cdap/api/dataset/lib/ObjectMappedTable.java (hunk 2 above)
Comment (anew) [inline cdap-api/src/main/java/co/cask/cdap/api/dataset/lib/ObjectMappedTable.java]: ok

File: cdap-api/src/main/java/co/cask/cdap/api/dataset/lib/ObjectStores.java (hunk 5 above)
Comment (anew) [inline cdap-api/src/main/java/co/cask/cdap/api/dataset/lib/ObjectStores.java]: cool

File: cdap-data-fabric/src/main/java/co/cask/cdap/data2/dataset2/lib/table/ObjectMappedTableDefinition.java (hunk 6 above)
Comment (anew) [inline cdap-data-fabric/src/main/java/co/cask/cdap/data2/dataset2/lib/table/ObjectMappedTableDefinition.java]: ok

File: cdap-data-fabric/src/main/java/co/cask/cdap/data2/dataset2/lib/table/ObjectMappedTableDataset.java (hunk 7 above)
Comment (anew) [inline cdap-data-fabric/src/main/java/co/cask/cdap/data2/dataset2/lib/table/ObjectMappedTableDataset.java]: ok

**Comment by anew** [comment]
only one more comment - I think you can bring back the write(byte[], T). Other than that - LGTM
